### PR TITLE
ENH: add rodent-specific adaption for B-spline

### PR DESCRIPTION
**Pull request by eilidhmacnicol**
Addressing #611, this PR 1) adds a function to determine bspline grid using a ratio of the number of slices in each dimension, and 2) adds logic to various workflows so that this might be used with rodents by updating the N4 parameters.  Not to say that the function can't be used with non-rodent images, but it hasn't been tested as such. 

There are other workflows that call `N4BiasFieldCorrection`, but they are unlikely to be used by NiRodents/fprodents because there are already rodent alternatives or because they would require more changes and would probably benefit from a rodent-specific workflow.

### CODE REVIEW

File: niworkflows/utils/images.py [hunk 1]
```diff
@@ -297,3 +297,20 @@ def nii_ones_like(in_file, value, dtype, newpath=None):
     nii.to_filename(out_file)
 
     return out_file
+
+
+def _bspline_grid(in_file):
```
Comment (eilidhmacnicol) [inline niworkflows/utils/images.py:302]: Inspiration for using slice number to determine grid sparsity: 
https://itk.org/pipermail/community/2014-February/005036.html

File: niworkflows/anat/ants.py [hunk 2]
```diff
@@ -863,6 +886,8 @@ def init_n4_only_wf(
     atropos_refine=True,
     atropos_use_random_seed=True,
     bids_suffix="T1w",
+    use_bspline_grid=False,
+    min_n4_iter=False,
```
Comment (oesteban) [inline niworkflows/anat/ants.py:890]: Do we use this workflow in nirodents? I don't think so, and I'd honestly avoid changing it at this moment.

File: niworkflows/utils/images.py (hunk 1 above)
Comment (oesteban) [reply niworkflows/utils/images.py:302]: Please add the link to the function's documentation

File: niworkflows/utils/images.py [hunk 3]
```diff
@@ -297,3 +297,20 @@ def nii_ones_like(in_file, value, dtype, newpath=None):
...
+    """Estimate B-Spline fitting distance grid using the number of slices of ``in_file``."""
+    import nibabel as nb
+    import numpy as np
+    import math
+
+    img = nb.load(in_file)
+    zooms = img.header.get_zooms()[:3]
+    # find extent of each dimension wrt voxel sizes
+    extent = (np.array(img.shape[:3]) - 1) * zooms
```
Comment (oesteban) [inline niworkflows/utils/images.py:311]:
```suggestion
    extent = np.array(img.shape[:3]) * zooms
```

Comment (oesteban) [reply niworkflows/utils/images.py:311]: extent is clear in definition - doesn't suffer from the fencepost problem

**Review by oesteban — COMMENTED** [review]
A few suggestions

**Comment by eilidhmacnicol** [comment]
Okay, tested the N4 in the reference workflow on two sets of multi-run EPIs from a subject, and it works well.
Although I was going to develop some tests for the `bspline_grid` function, I would quite like to push on with integrating the reference workflow into fmriprep-rodents. 

@oesteban what do you think?

**Comment by eilidhmacnicol** [comment]
`test_pytest` seems to be failing because of the pandas versioning rather than anything to do with the code, FYI

**Comment by oesteban** [comment]
I'll have a look ASAP

File: niworkflows/anat/ants.py [hunk 4]
```diff
@@ -735,15 +743,14 @@ def init_atropos_wf(
...
-            n_iterations=[50] * 5,
             convergence_threshold=1e-7,
             shrink_factor=4,
-            bspline_fitting_distance=bspline_fitting_distance,
         ),
         n_procs=omp_nthreads,
         name="inu_n4_final",
         iterfield=["input_image"],
     )
+    inu_n4_final.inputs.n_iterations = [50] * n4_iter
```
Comment (oesteban) [inline niworkflows/anat/ants.py:753]: why split this setting, instead of writing it above?

File: niworkflows/anat/ants.py [hunk 5]
```diff
@@ -855,6 +862,22 @@ def _argmax(in_dice):
...
+
+        # fmt:off
+        wf.connect([
+            (inputnode, bspline_grid, [(("in_files", _pop), "in_file")]),
+            (bspline_grid, inu_n4_final, [("out", "args")])
+        ])
+        # fmt:on
+    else:
+        # set INU bspline grid based on isotropic distance
+        inu_n4_final.inputs.bspline_fitting_distance = bspline_fitting_distance
```
Comment (oesteban) [inline niworkflows/anat/ants.py:879]:
```suggestion
```

You can restore line 741 to pick `bspline_fitting_distance` as default, and the connection will override the value when `adaptive_bspline_grid` is true.

File: niworkflows/utils/images.py [hunk 6]
```diff
@@ -297,3 +297,21 @@ def nii_ones_like(in_file, value, dtype, newpath=None):
     nii.to_filename(out_file)
 
     return out_file
+
+
+def _bspline_grid(in_file):
+    """Estimate B-Spline fitting distance grid using the number of slices of ``in_file``.
```
Comment (oesteban) [inline niworkflows/utils/images.py:303]:
```suggestion
    """
    Estimate B-Spline fitting distance grid using the number of slices of ``in_file``.

```

File: niworkflows/utils/images.py [hunk 7]
```diff
@@ -297,3 +297,21 @@ def nii_ones_like(in_file, value, dtype, newpath=None):
...
+    Using slice number to determine grid sparsity is inspired by the conversation found at
+    https://itk.org/pipermail/community/2014-February/005036.html"""
```
Comment (oesteban) [inline niworkflows/utils/images.py:305]:
```suggestion
    https://itk.org/pipermail/community/2014-February/005036.html
    
    """
```

File: niworkflows/workflows/epi/refmap.py [hunk 8]
```diff
@@ -64,6 +66,12 @@ def init_epi_reference_wf(
         If ``True``, determines nonsteady states in the beginning of the timeseries
         and selects them for the averaging of each run.
         IMPORTANT: this option applies only to BOLD EPIs.
+    adaptive_bspline_grid : :obj:`bool`
+        If ``True``, determines the number of b-spline grid elements from data shape
+        and feeds them into N4BiasFieldCorrection, rather than setting an isotropic distance.
+    n4_iter : :obj:`int`
+        The number of b-spline iterations (default: 5). Fewer (e.g. 4) are recommended
```
Comment (oesteban) [inline niworkflows/workflows/epi/refmap.py:73]:
```suggestion
        The number of B-Spline fitting iterations (default: 5). Fewer (e.g. 4) are recommended
```

File: niworkflows/workflows/epi/refmap.py [hunk 9]
```diff
@@ -64,6 +66,12 @@ def init_epi_reference_wf(
         If ``True``, determines nonsteady states in the beginning of the timeseries
         and selects them for the averaging of each run.
         IMPORTANT: this option applies only to BOLD EPIs.
+    adaptive_bspline_grid : :obj:`bool`
+        If ``True``, determines the number of b-spline grid elements from data shape
```
Comment (oesteban) [inline niworkflows/workflows/epi/refmap.py:70]:
```suggestion
        If ``True``, determines the number of B-Spline grid elements from data shape
```

File: niworkflows/workflows/epi/refmap.py [hunk 10]
```diff
@@ -202,6 +210,21 @@ def _set_threads(in_list, maximum):
     else:
         wf.connect(inputnode, "t_masks", per_run_avgs, "t_mask")
 
+    # rodent-specific N4 settings
+    n4_avgs.inputs.n_iterations = [50] * n4_iter
```
Comment (oesteban) [inline niworkflows/workflows/epi/refmap.py:214]: as above, why not setting on line 138?

File: niworkflows/anat/ants.py [hunk 11]
```diff
@@ -536,6 +538,12 @@ def init_atropos_wf(
         ``(4,4,2,3)`` uses K=4, CSF=4, GM=2, WM=3.
     bspline_fitting_distance : float
         The size of the b-spline mesh grid elements, in mm (default: 200)
+    adaptive_bspline_grid : :obj:`bool`
+        If true, defines the number of b-spline mesh grid elements in each dimension rather
```
Comment (oesteban) [inline niworkflows/anat/ants.py:542]:
```suggestion
        If true, defines the number of B-Spline mesh grid elements in each dimension rather
```

File: niworkflows/anat/ants.py [hunk 12]
```diff
@@ -536,6 +538,12 @@ def init_atropos_wf(
...
+        than using the isotropic distance given in ``bspline_fitting_distance``.
+    n4_iter : :obj:`int`
+        The number of b-spline iterations (default: 5). Fewer (e.g. 4) are recommended
```
Comment (oesteban) [inline niworkflows/anat/ants.py:545]:
```suggestion
        The number of B-Spline fitting iterations (default: 5). Fewer (e.g. 4) are recommended
```

**Review by oesteban — APPROVED** [review]
Looking good - left a few minor comments and nit picks

**Comment by oesteban** [comment]
> `test_pytest` seems to be failing because of the pandas versioning rather than anything to do with the code, FYI

I just rebased your branch on `upstream/master` where the problem has been recently addressed.

**Comment by oesteban** [comment]
I also have changed the target branch to `feat/infants-rodents` so that we don't mess up the release of fMRIPrep 21.0.x

**Comment by eilidhmacnicol** [comment]
@oesteban seems like the issue that was causing the tests to fail was related to the nipype version. Pinning to master has bypassed this, so is this ready to merge or should we pin to a newer version of nipype first?

**Comment by eilidhmacnicol** [comment]
More fun - it seems that whatever is causing the tests to fail has been resolved somewhere between the release of nipype 1.7.0 (https://github.com/nipy/nipype/tree/b38572030f3d968c945675efaccd053d123f158b; released 28 days ago at time of writing) and the version accessed today (https://github.com/nipy/nipype/tree/f756b71117493caea093bac79d576207d8f96482). 

My guess is that it is fixed by nipy/nipype#3395, so I guess we will have to pin to 1.7.1 when it is released.

**Comment by eilidhmacnicol** [comment]
In the meantime, I think I will revert to pinning to master and push through so we are not waiting solely on a nipype release

**Pull request merged**